be/src/storage/lake/txn_log_applier.cpp

```diff
@@ -49,6 +49,8 @@ class PrimaryKeyTxnLogApplier : public TxnLogApplier {
     ~PrimaryKeyTxnLogApplier() override {
         // must release primary index before `handle_failure`, otherwise `handle_failure` will fail
         _tablet.update_mgr()->release_primary_index_cache(_index_entry);
+        auto& index = _index_entry->value();
+        index.unlock();
```

Comment:
If `prepare_primary_index` fail, and we don't get lock yet, we don't need to unlock, right?